### PR TITLE
feat(groupChatManager): create a groupChat manager to create separate collections on a thread

### PR DESCRIPTION
**Pull request by Jekrimo**
**What this PR does** 📖
create a groupChat manager to create separate collections on a thread, allow sending messages, listening to the collection for changes, eventEmitters for testing I presume

**Which issue(s) this PR fixes** 🔨

AP-660

**Special notes for reviewers** 🗒️
EDIT: Testable now. 3 Groups are hardcoded for testing. Users will show as unknown unless it's your user. There are also other bug's and features missing currently. Most have been identified in AP-732 -> AP-744

### CODE REVIEW

File: libraries/Textile/GroupChatManager.ts [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+  async createGroupConversation(
+    friendIdentifiers: Array<string>,
+  ): Promise<string> {
+    const newCollectionUUID = uuid()
+    await this.textile.client.newCollection(this.threadID, {
+      name: newCollectionUUID,
+      schema: groupChatSchema,
+    })
+    // TODO: need somewhere to store new UUID for user to access
+    this.eventEmitter.emit(newCollectionUUID)
```
Comment (iltumio) [inline libraries/Textile/GroupChatManager.ts:81]: EventEmitter should extend the class, then you should be able to use the `this.emit()` method

File: libraries/Textile/GroupChatManager.ts [hunk 2]
```diff
@@ -0,0 +1,366 @@
+import { EventEmitter } from 'events'
```
Comment (iltumio) [inline libraries/Textile/GroupChatManager.ts:1]: Use the custome EventEmitter class we built for WebRTC that is strongly typed

**Review by iltumio — CHANGES_REQUESTED** [review]

**Comment by phillsatellite** [comment]
Tested: I'm not 100% sure if its caused by the PR but can't seem to get past the Linking Satellites page when creating a new account. I switched back to Dev branch, created new account and got in fine, switched back to this branch and created new account and still got blocked at Linking Satellites
Screen Shot 2022-02-03 at 2 20 40 PM (2)

**Review by phillsatellite — APPROVED** [review]

**Comment by stavares843** [comment]
tickets that jeff added

[AP-732](https://satellite-im.atlassian.net/browse/AP-732)
[AP-733](https://satellite-im.atlassian.net/browse/AP-733)
[AP-734](https://satellite-im.atlassian.net/browse/AP-734)
[AP-735](https://satellite-im.atlassian.net/browse/AP-735)
[AP-736](https://satellite-im.atlassian.net/browse/AP-736)
[AP-737](https://satellite-im.atlassian.net/browse/AP-737)
[AP-738](https://satellite-im.atlassian.net/browse/AP-738)
[AP-739](https://satellite-im.atlassian.net/browse/AP-739)
[AP-740](https://satellite-im.atlassian.net/browse/AP-740)
[AP-741](https://satellite-im.atlassian.net/browse/AP-741)
[AP-742](https://satellite-im.atlassian.net/browse/AP-742)
[AP-743](https://satellite-im.atlassian.net/browse/AP-743)
[AP-744](https://satellite-im.atlassian.net/browse/AP-744)

thanks for adding @Jekrimo 🔨

**Comment by phillsatellite** [comment]
Tested: Only issues I've found that I don't think weren't already addressed was the error in terminal (switched to branch, yarn, yarn dev to get that error) and the other error in Console "Error: errors.textile.friend_not_found"
Screen Shot 2022-02-09 at 4 49 06 PM

Screen Shot 2022-02-09 at 4 41 09 PM

**Comment by Jekrimo** [comment]
> Tested: Only issues I've found that I don't think weren't already addressed was the error in terminal (switched to branch, yarn, yarn dev to get that error) and the other error in Console "Error: errors.textile.friend_not_found" Screen Shot 2022-02-09 at 4 49 06 PM
> 
> Screen Shot 2022-02-09 at 4 41 09 PM

I think this has to do with us tricking it currently to allow multi users. I say we leave it and open a bug for it. But Idk what everyone else thinks

**Comment by stavares843** [comment]
do we know why appears the errors.textile.friend_not_found from Phil's screenshot? I see that added here
https://github.com/Satellite-im/locales/blob/main/en-US.js#L469


can we fix the warning when it finishes compiling locally? atm we have zero compiling warnings after compiling

**Comment by stavares843** [comment]
is possible to get the textile error that phil mentioned above fixed as well? 🔨

**Comment by Jekrimo** [comment]
I'm tryin to recreate it and can't get the error, is it showing for anyone else?

**Comment by phillsatellite** [comment]
> I'm tryin to recreate it and can't get the error, is it showing for anyone else?

Hey Jeff! Sorry didn't notice earlier I got that error from trying to react to a message, which you said above isnt working yet 👍

**Comment by stavares843** [comment]
thanks for testing @phillsatellite

**Comment by Jekrimo** [comment]
Awesome! That's great news, perfect!

File: components/views/chat/chatbar/Chatbar.vue [hunk 3]
```diff
@@ -271,10 +271,22 @@ export default Vue.extend({
...
-        this.$store.dispatch('textile/sendTextMessage', {
-          to: this.recipient.textilePubkey,
-          text: value,
-        })
+
+        // Check if it's a group
+        if (
+          RegExp(this.$Config.regex.uuidv4).test(this.recipient.textilePubkey)
+        ) {
+          this.$store.dispatch('textile/sendGroupMessage', {
```
Comment (WanderingHogan) [inline components/views/chat/chatbar/Chatbar.vue:279]: what do you think about dumping the if/else and doing something like 

```
        // Check if the message is a group or a direct message
        const messageType = RegExp(this.$Config.regex.uuidv4).test(
          this.recipient.textilePubkey,
        )
          ? 'sendGroupMessage'
          : 'sendTextMessage'

        this.$store.dispatch(`textile/${messageType}`, {
          to: this.recipient.textilePubkey,
          text: value,
        })
        ```

File: components/views/chat/conversation/Conversation.vue [hunk 4]
```diff
@@ -4,6 +4,7 @@ import Vue, { PropType } from 'vue'
 import { mapState } from 'vuex'
 
 import { MessageGroup } from '~/types/messaging'
+import { Message } from '~/types/textile/mailbox'
```
Comment (WanderingHogan) [inline components/views/chat/conversation/Conversation.vue:7]: is this getting ready for something in the future, or leftover

File: layouts/chat.vue [hunk 5]
```diff
@@ -147,7 +147,11 @@ export default Vue.extend({
       const isMe =
         this.$route.params.address === this.$typedStore.state.accounts.active
 
-      const recipient = isMe
+      const groupId = this.$route.params.id
+
+      const recipient = groupId
+        ? { textilePubkey: groupId }
+        : isMe
```
Comment (WanderingHogan) [inline layouts/chat.vue:154]: How is this meant to work?

Comment (WanderingHogan) [reply layouts/chat.vue:154]: is it missing something to wrap that first ternary, or does it not need it?

Comment (WanderingHogan) [reply layouts/chat.vue:154]: TIL: conditional chains in ternary operators

**Review by WanderingHogan — APPROVED** [review]

**Pull request merged**